File: java/src/main/java/examples/UimIntegrationUtils.java
package examples;

import org.json.JSONObject;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Base64;

public class UimIntegrationUtils {
    private String clientIdentifier;
    private String apiKey;
    private String environment;

    public UimIntegrationUtils(String clientIdentifier, String apiKey, String environment) {
        this.clientIdentifier = clientIdentifier;
        this.apiKey = apiKey;
        this.environment = environment;
    }

    public String sendSignedRequest(String method, String requestPath, String query, String body) throws Exception {
        return sendSignedRequest(method, requestPath, query, body, false);
    }

    public String sendSignedRequest(String method, String requestPath, String query, String body, boolean useFormAuth) throws Exception {
        String acceptHeader = "application/vnd.rb.uim-v1+json";
        String date = currentDatetime();
        String base64Body = body != null ? Base64.getEncoder().encodeToString(body.getBytes()): "";
        String signature = createSignature(method, requestPath, query, acceptHeader, base64Body, date);
        String url = uimUrl(requestPath);
        if(query != null && !query.isEmpty()) {
            url += "?" + query;
        }
        return callUim(method, body, signature, url, acceptHeader, date, useFormAuth);
    }

    private String createSignature(String requestMethod, String requestPath, String query, String acceptHeader, String body, String date) {
        String[] params = {clientIdentifier, requestMethod, requestPath, query, acceptHeader, body, date};
        String rawSignature = String.join("\n", params);

        try {
            Mac sha256HMAC = Mac.getInstance("HmacSHA256");
            SecretKeySpec secretKey = new SecretKeySpec(apiKey.getBytes(), "HmacSHA256");
            sha256HMAC.init(secretKey);

            String hash = Base64.getEncoder().encodeToString(sha256HMAC.doFinal(rawSignature.getBytes()));
            return hash;
        }
        catch (Exception e){
            System.err.println("Error");
        }
        return "";
    }

    private String uimUrl(String requestPath) {
        String baseUrl;
        if (environment.equals("design")) {
            baseUrl = "https://uim-design.redbull.com/uim";
        } else if (environment.equals("production")) {
            baseUrl = "https://uim.redbull.com/uim";
        } else if (environment.equals("local")) {
            baseUrl = "http://localhost:8090";
        } else {
            throw new Error("Unknown environment " + environment);
        }
        return baseUrl + requestPath;
    }

    private String createHmac(String signature, boolean useFormAuth) {
        String prefix = useFormAuth ? "FORM:" : "";
        String hmacSignature = "HMAC " + prefix + clientIdentifier + ":" + signature;
        System.out.println("Client: " + clientIdentifier + "\n");
        System.out.println("Signature: " + hmacSignature + "\n");
        return hmacSignature;
    }

    private String callUim(String method, String body, String signature, String url, String acceptHeader, String date, boolean useFormAuth) throws Exception {
        String hmacSignature = createHmac(signature, useFormAuth);
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Content-Type", "application/json; charset=utf-8")
                .header("Accept", acceptHeader)
                .header("UIM-Date", date)
                .header("Authorization", hmacSignature);
        if ("POST".equals(method)) {
            requestBuilder = requestBuilder.POST(HttpRequest.BodyPublishers.ofString(body));
        } else if ("GET".equals(method)) {
            requestBuilder = requestBuilder.GET();
        }
        HttpResponse<String> response = client.send(requestBuilder.build(),
                HttpResponse.BodyHandlers.ofString());

        System.out.println(response.body());
        return response.body();
    }

    private String currentDatetime() {
        return ZonedDateTime.now(ZoneId.of("GMT")).format(DateTimeFormatter.RFC_1123_DATE_TIME);
    }

    public JSONObject getRegistrationItems() throws Exception {
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(uimUrl("/api/application/registration-items")))
                .header("Application-Id", clientIdentifier)
                .header("Accept", "application/vnd.rb.uim-v16+json")
                .GET()
                .build();

        HttpResponse<String> response = client.send(request,
                HttpResponse.BodyHandlers.ofString());

        JSONObject registrationItems = new JSONObject(response.body());
        return registrationItems;
    }
}

File: csharp/UimFindUser/UimIntegrationUtils.cs
using System;
using System.Text;
using System.IO;
using System.Security.Cryptography;
using System.Net;

namespace Utils {
    class UimIntegrationUtils {

        public UimIntegrationUtils(string appId, string apiKey, string environment) {
            ClientIdentifier = appId;
            ApiKey = apiKey;
            Environment = environment;
        }

        public string ClientIdentifier { get; }
        public string ApiKey { get; }
        public string Environment { get; }

        private string currentDatetime() {
            return DateTime.Now.ToUniversalTime().ToString("r");
        }

        private string base64(string payload) {
            var plainTextBytes = System.Text.Encoding.UTF8.GetBytes(payload);
            return System.Convert.ToBase64String(plainTextBytes);
        }

        private string hmacsha256(string key, string text) {
            UTF8Encoding encoding = new System.Text.UTF8Encoding();
            byte[] keyBytes = encoding.GetBytes(key);
            byte[] messageBytes = encoding.GetBytes(text);
            using (var hmacsha256 = new HMACSHA256(keyBytes))
            {
                byte[] hashmessage = hmacsha256.ComputeHash(messageBytes);
                return System.Convert.ToBase64String(hashmessage);
            }

        }

        private string createSignature(string requestMethod, string requestPath, string query, string acceptHeader, string body, string date) {
            String base64Body = body != null ? base64(body) : "";
            String rawSignature = ClientIdentifier + "\n" +
                                 requestMethod + "\n" +
                                 requestPath + "\n" +
                                 query + "\n" +
                                 acceptHeader + "\n" +
                                 base64Body + "\n" +
                                 date;
            return hmacsha256(ApiKey, rawSignature);
        }

        private string uimUrl(string requestPath) {
            string baseUrl;
            if (Environment.Equals("design")) {
                baseUrl = "https://uim-design.redbull.com/uim";
            } else if (Environment.Equals("production")) {
                baseUrl = "https://uim.redbull.com/uim";
            } else if (Environment.Equals("local")) {
                baseUrl = "http://localhost:8090";
            } else {
                throw new System.Exception("Unknown environment " + Environment);
            }
            return baseUrl + requestPath;
        }

        private string callUim(string method, string body, string signature, string url, string acceptHeader, string date, bool useFormAuth) {
            WebClient webClient = new WebClient();
            webClient.Headers.Add("Accept", acceptHeader);
            webClient.Headers.Add("Content-Type", "application/json; charset=utf-8");
            webClient.Headers.Add("UIM-Date", date);
            webClient.Headers.Add("Authorization", "HMAC " + ClientIdentifier + ":" + signature);
            string response = "empty";
            if ("GET".Equals(method)) {
                response = webClient.DownloadString(url);
            }
            return response;
        }

        public string sendSignedRequest(string method, string requestPath, string query, string body) {
            return sendSignedRequest(method, requestPath, query, body, false);
        }

        public string sendSignedRequest(string method, string requestPath, string query, string body, bool useFormAuth) {
            string acceptHeader = "application/vnd.rb.uim-v1+json";
            string date = currentDatetime();
            string signature = createSignature(method, requestPath, query, acceptHeader, body, date);
            string url = uimUrl(requestPath);
            if (!String.IsNullOrEmpty(query)) {
                url += "?" + query;
            }
            return callUim(method, body, signature, url, acceptHeader, date, useFormAuth);
        }
    }
}